### PR TITLE
OCS Fixes to allow setting of password without removing additional settings

### PR DESCRIPTION
**Pull request by rullzer**
Redo of #14868 but only for the OCS API.

What has been done:
- Added setPassword to share.php
- Fixed OCS API call
- Added unit tests

Related issues #10671 and #14826 

No work has been done in the web frontend since we need to switch that to the OCS API anyway. And the patch did not fix it completely anyway.

Might be a good idea to backport this fix as well. Since the clients use the OCS API.

### CODE REVIEW

**Comment by rullzer** [comment]
@owncloud-bot retest this please

**Comment by rullzer** [comment]
CC @schiesbn @PVince81 @LukasReschke @DeepDiver1975

File: lib/public/idbconnection.php [hunk 1]
```diff
@@ -167,4 +167,11 @@ public function dropTable($table);
 	 * @return bool
 	 */
 	public function tableExists($table);
+
+	/**
+	 * Gets the QueryBuilder
+	 *
+	 * @return \Doctrine\DBAL\Query\QueryBuilder
+	 */
+	public function createQueryBuilder();
```
Comment (DeepDiver1975) [inline lib/public/idbconnection.php]: please don't expose that one on the public api. 
I'd rather have our own implementation/wrapper for the query builder which takes care about the prefix and the backquotes. Once we have that querybuilder we can add the public api.

THX

File: lib/private/share/share.php [hunk 2]
```diff
@@ -1123,6 +1127,49 @@ public static function setExpirationDate($itemType, $itemSource, $date, $shareTi
...
+	 * @param IUserSession $userSession
+	 * @param IDBConnection $connection
+	 * @param IConfig $config
+	 * @param string $itemType
+	 * @param string $itemSource
+	 * @param string $password
+	 * @throws \Exception
+	 * @return boolean
+	 */
+	public static function setPassword(IUserSession $userSession,
```
Comment (DeepDiver1975) [inline lib/private/share/share.php]: don't we have the share id at hand meanwhile?

the signature should be something like setPassword($id, $password)

**Comment by DeepDiver1975** [comment]
@rullzer rebase please

File: lib/private/share/share.php (hunk 2 above)
Comment (rullzer) [inline lib/private/share/share.php]: Yes we do. Should look much nicer now.

**Comment by rullzer** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
@owncloud-bot retest this please

**Comment by rullzer** [comment]
@owncloud-bot retest this bot

**Comment by DeepDiver1975** [comment]
@owncloud-bot retest this bot

**Comment by rullzer** [comment]
@owncloud-bot retest this please

**Comment by rullzer** [comment]
Jenkins PR in #15068

File: lib/private/share/share.php [hunk 3]
```diff
@@ -1131,6 +1135,44 @@ public static function setExpirationDate($itemType, $itemSource, $date, $shareTi
...
+	 */
+	public static function setPassword(IUserSession $userSession, 
+	                                   Connection $connection,
+	                                   IConfig $config,
+									   $shareId, $password) {
+		$user = $userSession->getUser();
+		if (is_null($user)) {
+			throw new \Exception("User not logged in");
+		}
+
```
Comment (rullzer) [inline lib/private/share/share.php]: mmm, we still need to validate that the share we update is actually our share. Else bad apps might do bad things!

**Comment by rullzer** [comment]
@owncloud-bot retest this please

**Comment by LukasReschke** [comment]
```
10:12:23 ............................................PHP Fatal error:  Call to a member function from() on null in /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/share/share.php on line 1148
```

**Comment by rullzer** [comment]
@LukasReschke ah yes that is solved. Problem now lies in something in the oracle setup (I have to setup docker to debug) see #15137

**Comment by DeepDiver1975** [comment]
@rullzer rebase please

**Comment by DeepDiver1975** [comment]
@rullzer still developing or ready to review? THX

**Comment by rullzer** [comment]
Jenkins is happy see #15137
I rebased and squased the commits.
Lets do this :)

**Comment by rullzer** [comment]
@DeepDiver1975 @LukasReschke @schiesbn @PVince81 I think it would be good to get this into 8.1 as well.

File: lib/private/share/share.php [hunk 4]
```diff
@@ -1151,6 +1155,74 @@ public static function setExpirationDate($itemType, $itemSource, $date, $shareTi
 	}
 
 	/**
+	 * Retrieve the owner of a connection
+	 *
+	 * @param Connection $connection
```
Comment (LukasReschke) [inline lib/private/share/share.php]: IDBConnection?

File: lib/private/share/share.php [hunk 5]
```diff
@@ -1151,6 +1155,74 @@ public static function setExpirationDate($itemType, $itemSource, $date, $shareTi
...
+	 * @param int $shareId
+	 * @throws \Exception
+	 * @return string uid of share owner
+	 */
+	private static function getShareOwner(Connection $connection, $shareId) {
+		$qb = $connection->createQueryBuilder();
+
+		$qb->select('`uid_owner`')
+			->from('`*PREFIX*share`')
+			->where($qb->expr()->eq('`id`', $shareId));
```
Comment (LukasReschke) [inline lib/private/share/share.php]: :bomb: :bomb: :bomb:

The DBAL query builder requires also prepared statements or you're getting yourself in SQL Injection problems.

Please take a look at http://doctrine-dbal.readthedocs.org/en/latest/reference/query-builder.html#security-safely-preventing-sql-injection

@DeepDiver1975 FYI - as we discussed this recently - we need to watch this very carefully.

:bomb: :bomb: :bomb:

File: lib/private/share/share.php [hunk 6]
```diff
@@ -1151,6 +1155,74 @@ public static function setExpirationDate($itemType, $itemSource, $date, $shareTi
...
+
+		//If passwords are enforced the password can't be null
+		if (self::enforcePassword($config) && is_null($password)) {
+			throw new \Exception('Cannot remove password');
+		}
+
+		$qb = $connection->createQueryBuilder();
+		$qb->update('`*PREFIX*share`')
+			->set('`share_with`', is_null($password) ? 'NULL' : $qb->expr()->literal(\OC::$server->getHasher()->hash($password)))
+			->where($qb->expr()->eq('`id`', $shareId));
```
Comment (LukasReschke) [inline lib/private/share/share.php]: :bomb: :bomb: :bomb:

The DBAL query builder requires also prepared statements or you're getting yourself in SQL Injection problems.

Please take a look at http://doctrine-dbal.readthedocs.org/en/latest/reference/query-builder.html#security-safely-preventing-sql-injection

@DeepDiver1975 FYI - as we discussed this recently - we need to watch this very carefully.

:bomb: :bomb: :bomb:

File: lib/private/share/share.php [hunk 7]
```diff
@@ -1151,6 +1155,74 @@ public static function setExpirationDate($itemType, $itemSource, $date, $shareTi
...
+		}
+
+		//If passwords are enforced the password can't be null
+		if (self::enforcePassword($config) && is_null($password)) {
+			throw new \Exception('Cannot remove password');
+		}
+
+		$qb = $connection->createQueryBuilder();
+		$qb->update('`*PREFIX*share`')
+			->set('`share_with`', is_null($password) ? 'NULL' : $qb->expr()->literal(\OC::$server->getHasher()->hash($password)))
```
Comment (LukasReschke) [inline lib/private/share/share.php]: Also try here whether prepared statements are possible

**Comment by LukasReschke** [comment]
Works :+1:

**Comment by MorrisJobke** [comment]
Tested and works :+1:

**Comment by PVince81** [comment]
CC @SergioBertolinSG you should be able to update the password properly on master now through OCS, can you check it ?

**Comment by SergioBertolinSG** [comment]
I have tested #14826, I can't add a password correctly using
 curl -v -X PUT -d password=sergio http://USER:PASSWORD@OCPATH/ocs/v1.php/apps/files_sharing/api/v1/shares/SHARE_ID

It puts a password, but it is rejected when i open the shared link.

Also putting a password using the web ui makes the expiration date dissapear, the UI calls this code?

**Comment by PVince81** [comment]
@SergioBertolinSG this only fixes the OCS call. The web UI's share dropdown is not using OCS (but should in the future)

**Pull request merged**